architecture/traffic-analysis-mitigation-with-decoys.md

```diff
@@ -0,0 +1,257 @@
+
+* The first call (to /register) might fail due to communications failure or server failure, resulting in one or even more retries.
+
+* The second call (to /postkeys) might fail due to communications failure or server failure, resulting in one or even more retries.
+
+* The user may change his mind about uploading his keys after the /register call has been made. Then the /postkeys call will not be invoked.
+
+# Equal request and response message lengths
+
+An observer can detect and analyse all the Txxx_xxx request and response times, and all the Lxxx_xxx message lengths in the above table.
+
+A first measure to mitigate traffic analysis is:
+
+All calls to /register and /postkeys will have equal request and response sizes, by using padding.
+
+* Lreg_req must be equal to Lpst_req
+
+* Lreg_rsp must be equal to Lpst_rsp.
+
+This requires both the apps (Android and iOS) and the back end server to agree upon the request and response message lengths for the /register and /postkey implementations, so that the request and response payload will always fit.
+
+As a consequence of this measure, an observer cannot distinguish between /register calls and /postkeys calls.
+
+# TEK upload observation
+
+Without any decoy traffic, an observer can deduce the following:
+
+* 0 calls on a day: no upload 
+
+* 1 call on a day and no call on day after: no upload
+
+* 1 call on a day and 1 call on following day: possible upload (/register may have been called twice; or genuine /register and a delayed /postkeys after Midnight)
+
+* 1 call on a day and 2 calls on following day: possible upload (/register may have been called twice; or /register, 1st /postkeys delayed until after Midnight, and another /register)
+
+* 2 calls on a day with interval < 5 min: very likely upload (/register and /postkeys have been invoked, or /register has been retried, which is unlikely)
+
+* 2 calls on a day with interval > 5 min: very likely upload (/register and /postkeys have been invoked, and for some reason the user opened upload screen twice or more that day.)
+
+* more than 2 calls in a day: very likely upload (/register and/or /postkeys has been retried. /postkeys might have failed, but unlikely.)
+
+## Combination of TEK upload and phone call
+
+Another datum that may be observed is that during a genuine TEK upload (/register and /postkeys calls) is having a phone call with the health care worker. If this is a VoIP call, this will be visible on the WiFi network. If this is a cell phone call, its call metadata could be analysed by e.g. corporate device management software, or by a nearby observer.
+
+Decoy traffic as described below will make this analysis less valuable.
+
+# Decoy traffic design
+
+We take two measures to introduce decoy traffic:
+
+1. Generate decoy traffic when user opens upload screen, to make genuine and simulated TEK uploads indistinguishable
+
+2. Generate simulated TEK uploads on random days.
```

Comment:
Two questions:
1. Will this be a background activity or does it require the user to open the key upload screen?
2. If this only happens when the user opens the key upload screen, when will the user open this screen?

Reply:
This will be a scheduled background activity. So user does not need to open the screen.